Vhar 7376 ilmoitukset raportointi

solita-staskila opened the pull request.

solita-staskila committed:
Raportin pohjakoodi

solita-staskila committed:
Merge branch 'develop' into VHAR-7376_ilmoitukset_raportointi

solita-staskila committed:
Excel&pdf korjauksia, taulukot ja niiden parametrit lisätty

solita-staskila committed:
UI korjaukset & testi

solita-staskila committed:
Merge branch 'develop' into VHAR-7376_ilmoitukset_raportointi

solita-staskila commented:
Vaatii jonkun rajoituksen, testattu singletonilla, jää roikkumaan jos ilmoituksia auki 500 ja painaa tallenna excel. 
Mahdollisesti kaatuu koko järjestelmä.

solita-staskila committed:
Tallenna- napin tyyli

solita-staskila committed:
Pixelin viilausta

solita-staskila committed:
Roolit korjattu, excelin lisätieto-sarakkeeseen lisätty rivitys

solita-staskila committed:
Merge branch 'develop' into VHAR-7376_ilmoitukset_raportointi

jpluomajoki-solita commented on dev-resources/less/pages/ilmoitukset.less:
```diff
@@ -203,15 +203,15 @@
 
 .TPP {
-  color: @pinkki-kevyt;
+  color: @pinkki-kevyt !important;
```
Ovatko nämä tarpeellisia? Jos ei, niin pois.

solita-staskila replied:
Ei ole tarpeellisia. 
https://github.com/finnishtransportagency/harja/pull/2824/commits/abaaadd87e60f4905bea28ad0040d7a019ccb4a9

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,200 @@
+  (let [ilmoitustyyppi (domain/ilmoitustyypin-lyhenne (:ilmoitustyyppi tiedot))
+        tietojen-selitteet-koko (count (:selitteet tiedot))
+        ;; Rivittää lisätieto-kentän joka 90. kirjaimen seuraavalle riville, jos lisätiedossa paljon tekstiä 
+        lisatieto-rivitetty (clojure.string/replace (str (:lisatieto tiedot)) #"(.{90})(?!$)" "$1\n")
```
clojure.stringin voisi requirata, tai poistaa tämän rivin kokonaan, kts. toinen kommentti.

solita-staskila replied:
Lisätty require
https://github.com/finnishtransportagency/harja/pull/2824/commits/abaaadd87e60f4905bea28ad0040d7a019ccb4a9

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,200 @@
+  (let [ilmoitustyyppi (domain/ilmoitustyypin-lyhenne (:ilmoitustyyppi tiedot))
+        tietojen-selitteet-koko (count (:selitteet tiedot))
+        ;; Rivittää lisätieto-kentän joka 90. kirjaimen seuraavalle riville, jos lisätiedossa paljon tekstiä 
+        lisatieto-rivitetty (clojure.string/replace (str (:lisatieto tiedot)) #"(.{90})(?!$)" "$1\n")
```
Sanan wräppäys käsin voi olla hieman huono käytettävyydeltä. Parempi ratkaisu voisi olla asettamalla sarakkeen leveys ja asettamalla sen tyyleihin word wrap päälle. Ainakin excelissä pitäisi toimia tällainen, PDF:stä en osaa sanoa. Jos vaikuttaa liian työläältä niin saanee kelvata toistaiseksi.

solita-staskila replied:
Tämä tulee excel.clj taulukko komponenttiin jossa laitetaan automaattiset leveydet tekstin pituuden mukaan, pitäisi mahdollisesti lisätä taulukkoon uusi parametri(?) jotta saa exceliin tuon tehtyä. Tulee enemmän vaivaa tehdä tämä yhdelle sarakkeelle, minusta parempi ratkaisu vaan tuo manuaalinen sanan wräppäys. 

PDF puolelle tuota lisätieto saraketta voi leventää, levensin koska oli turhan pieni https://github.com/finnishtransportagency/harja/pull/2824/commits/abaaadd87e60f4905bea28ad0040d7a019ccb4a9#diff-1f6dfe084f075c0e0e3aab3543466e16422ba61a5d7d7a0d4db2912beab1c1feL49
siinä on myös tuo wräppäys mukana mutta se on ihan OK.

solita-staskila replied:
Tähän vielä, vois tuonne exceliin kyllä jonkin virvelöinnin tehdä tuohon kentän leveyteen, voisi olla tulevaisuuden kannalta parempi, mutta se voi tosiaan mennä vaikeaksi.

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,200 @@
+        lisatieto-rivitetty (clojure.string/replace (str (:lisatieto tiedot)) #"(.{90})(?!$)" "$1\n")
+        ;; Käydään läpi selitteet, lisää pilkun jos enemmän selitteitä olemassa
+        ;; "Savea tiellä, Vettä tiellä"
+        tietojen-selitteet (map (fn [x]
```
Katso `clojure.string/join` (ja muista requirata clojure.string, vaikkapa str-aliaksella, esim. `[clojure.string :as str]`

solita-staskila replied:
Korjattu https://github.com/finnishtransportagency/harja/pull/2824/commits/abaaadd87e60f4905bea28ad0040d7a019ccb4a9

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,200 @@
+                          {:otsikko x
+                           :leveys 36}) otsikot)]
+
+       ;; TODO en tiedä voiko tätä jotenkin nätimmin tehdä
```
Voi. `(apply rivi testets)`

jpluomajoki-solita replied:
Oikeastaan voi siistiä vielä lisää. 
`rivi`-funktiota ei tarvi käyttää. Se vain muuttaa parametrit vektoriksi ja ottaa nillit pois. Otsikot määritellään itse niin sitä ei tarvi käyttää.
Lisäksi otsikoita ei tarvi turhaan kuljettaa parametreissa, vaan ne voidaan määritellä vaikka suoraan tässä valmiiksi oikeassa formaatissa. `:leveys` ei taida tehdä myöskään mitään, joten senkin voi ottaa samalla pois.

solita-staskila replied:
> Voi. (apply rivi testets)

Tämä korjattu, + muuttujan nimi kun kissa olisi kävellyt näppäimistöllä, tämä myös vaihdettu https://github.com/finnishtransportagency/harja/pull/2824/commits/abaaadd87e60f4905bea28ad0040d7a019ccb4a9


> Oikeastaan voi siistiä vielä lisää.
> rivi-funktiota ei tarvi käyttää. Se vain muuttaa parametrit vektoriksi ja ottaa nillit pois. Otsikot määritellään itse niin sitä ei tarvi käyttää.
> Lisäksi otsikoita ei tarvi turhaan kuljettaa parametreissa, vaan ne voidaan määritellä vaikka suoraan tässä valmiiksi oikeassa formaatissa. :leveys ei taida tehdä myöskään mitään, joten senkin voi ottaa samalla pois.

En ymmärrä. Leveys määrittää rivien leveyden, otsikko määrittää rivien otsikon.
 Tässä rivin leveys 5: 
![image](https://user-images.githubusercontent.com/116802938/230107264-ad19c906-7d44-4d00-97ff-b5ac2975ddb2.png)

Tässä rivin :leveys 36: 
![image](https://user-images.githubusercontent.com/116802938/230107326-00508425-f652-407b-971b-f22a5a332f48.png)

Pitänee katsoa tämä parikoodauksella.

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,200 @@
+
+(defn- taulukko [{:keys [otsikot tiedot]}]
+  (let [rivit (into []
+                    (remove nil?
```
`(remove nil?)` taitaa olla turha, käsittääkseni `(rivi-taulukolle)` ei koskaan palauta nil, vaan vähintään tyhjän vektorin.

solita-staskila replied:
Joo, korjattu 
https://github.com/finnishtransportagency/harja/pull/2824/commits/37efc5b5cf81c4a5cd0d9c9dac6a8487fc9b5d62

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,200 @@
+(defn- taulukko [{:keys [otsikot tiedot]}]
+  (let [rivit (into []
+                    (remove nil?
+                            (for [x tiedot]
```
for -> map, `(map #(rivi-taulukolle % true) tiedot)`

jpluomajoki-solita replied:
Lisäksi, onko rivien aina tarkoitus olla lihavoituja?

solita-staskila replied:
Korjattu, ei oikeastaan tarvi tässä lihavointia ollenkaan.
https://github.com/finnishtransportagency/harja/pull/2824/commits/0af965568c0ef1976255904a335d83f0fcc7e9b7

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,200 @@
+                 "Lisätieto" "Tie"
+                 "Tila" "Toimenpiteet aloitettu"]
+
+        valittu-ely (try (get ely/elynumero->nimi (Long/parseLong (:elynumero hallintayksikko))) (catch Throwable _ nil))
```
try-catchin sijasta käytetään yleensä `(when (:elynumero hallintayksikko) (:elynumero hallintayksikko)) ` tms.
Liian aktiivinen try-catch, erityisesti näin laajalla catchilla voi ottaa joissain tilanteissa kiinni sellaisia tilanteita, jotka on oikeasti virheitä, jolloin oikea virhe jää helposti huomaamatta. Tässä tilanteessa luultavasti ei niin kävisi, mutta silti on hyvä välttää try-catchia kun voi.

solita-staskila replied:
Otettu try catch pois

https://github.com/finnishtransportagency/harja/pull/2824/commits/0af965568c0ef1976255904a335d83f0fcc7e9b7

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,200 @@
+        valittu-ely (try (get ely/elynumero->nimi (Long/parseLong (:elynumero hallintayksikko))) (catch Throwable _ nil))
+
+        otsikko (if-not (:nimi urakka)
+                  (str "Ilmoitus raportti, " valittu-ely) "Ilmoitus raportti")
```
Raportti-sana pois näistä, ei ole käyttäjän näkökulmasta raportti.

solita-staskila replied:
Korjattu 
https://github.com/finnishtransportagency/harja/commit/dd350019919cbb36347c9e4bed22b2f78375a70a

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/yleinen.clj:
```diff
@@ -234,7 +234,20 @@
   "Tietoja ei löytynyt valitulta aikaväliltä.")
 
 (defn raportti-tiedostonimi [raportin-tunnistetiedot]
-  (str/join ", "
```
Miksi tämä muutos on tehty?

solita-staskila replied:
Vanhalla koodilla oli mahdollista syntyä tiedostonimiä joissa tyhjiä pilkkuja kuten: 
"Raportti, ,uusimaa" tms. enkä ymmärtänyt vanhaa koodia, tein tästä hieman selkeämmän. 

Toimii siis nyt aivan samalla lailla kun ennenkin, mutta ei voi tulla tuommoisia tyhjiä pilkkuja tiedostonimiin.

jpluomajoki-solita replied:
Aivan, luin väärin ja luulin että tässä oli muutettu tiedostonimeä. Siinä tapauksessa tämä OK. Luulen, että tämän saisi toimimaan edelleenkin `str/join`illa, kunhan vaan siistitään nillit pois tuosta listasta, esim `(keep identity ((juxt...`

jpluomajoki-solita commented on src/cljs/harja/ui/raportti.cljs:
```diff
@@ -343,6 +348,11 @@
                              [sisalto]))
                          sisalto))])
 
+(defmethod muodosta-html :piilota-html [[_ & _]]
```
Tarvitaanko tämmöistä oikeasti?

solita-staskila replied:
Ei tarvita, korjattu https://github.com/finnishtransportagency/harja/commit/07fef416b724112923cb599d0ec927c8577d0999

jpluomajoki-solita reviewed (changes requested):
- "Raportti"-sanan voisi poistaa kokonaan excelin&pdf nimestä ja sisällöstä. Tämä ei ole käyttäjälle oikeastaan raportti, vaan lista ilmoituksista excel-muodossa. Sisäisesti voidaan kyllä käsitellä raporttina, mutta käyttäjälle se ei ole sitä.
- Raportin muodostusta kutsutaan tarpeettomasti. Aina kun urakka tai filtterit muuttuu, tehdään uusi raportti, mutta sitä ei kuitenkaan näytetä missään.
- Ymmärtääkseni tämä ei ole ensimmäinen raportti, josta tehdään vain excel/pdf, joten epäilen, että tuo :piilota-html voi olla turha. Ainakin `:kanavien-liikennetapahtumat` taitaa olla hyvin samanlainen kuin tämä, eli kälille haetaan eri tavalla ja sitten excel/pdf viedään "raporttina"

solita-staskila committed:
Raportti- sana poistettu käyttäjiltä

solita-staskila committed:
Korjattu raportointi kutsut & napin tyyli lisätty gridiin

solita-staskila committed:
Palautettu excel & pdf koodi, poistettu :piilota-html komponentti

solita-staskila commented:
> * "Raportti"-sanan voisi poistaa kokonaan excelin&pdf nimestä ja sisällöstä. Tämä ei ole käyttäjälle oikeastaan raportti, vaan lista ilmoituksista excel-muodossa. Sisäisesti voidaan kyllä käsitellä raporttina, mutta käyttäjälle se ei ole sitä.
> * Raportin muodostusta kutsutaan tarpeettomasti. Aina kun urakka tai filtterit muuttuu, tehdään uusi raportti, mutta sitä ei kuitenkaan näytetä missään.
> * Ymmärtääkseni tämä ei ole ensimmäinen raportti, josta tehdään vain excel/pdf, joten epäilen, että tuo :piilota-html voi olla turha. Ainakin `:kanavien-liikennetapahtumat` taitaa olla hyvin samanlainen kuin tämä, eli kälille haetaan eri tavalla ja sitten excel/pdf viedään "raporttina"

- Raportti maininta poistettu käyttäjiltä
https://github.com/finnishtransportagency/harja/pull/2824/commits/dd350019919cbb36347c9e4bed22b2f78375a70a

- Korjattu raportin kutstunta -> raportointi siirretty gridiin :raporttivienti parametreihin, ja poistettu turha cljs tiedosto
Raportti suoritetaan vain silloin kun painetaan tallenna- nappeja 
https://github.com/finnishtransportagency/harja/pull/2824/commits/2ebc768daa8c90723a7248ba959db24329abdd03

- Poistettu :piilota-html komponentti koodeineen kokonaan, tämä oli tosiaan turha
https://github.com/finnishtransportagency/harja/pull/2824/commits/07fef416b724112923cb599d0ec927c8577d0999

solita-staskila committed:
Turha CSS & siistitty koodia

solita-staskila committed:
Siistitty taulukon rivi koodi

solita-staskila committed:
Try-catch pois, lihavointi pois, mäppäys siistitty

solita-staskila committed:
Turha check

solita-staskila committed:
Merge branch 'develop' into VHAR-7376_ilmoitukset_raportointi

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/pdf.clj:
```diff
@@ -501,15 +505,21 @@
-             (str otsikko ":")))]]
-       [:fo:table-cell
-        [:fo:block {:margin-left "5mm"} (str arvo)]]])]])
+      (let [arvo (if (and
```
Oliko tätä testattu muissa raporteissa?

solita-staskila replied:
Joo, on testattu. 
Eli sen sijaan että Urakka / Urakoitsija - sarakkeet olisi vain tyhjiä, näissä nyt lukee "Kaikki" , tai urakoitsijan kohdalla "-" jos urakoitsijaa ei ole valittu. 
![image](https://user-images.githubusercontent.com/116802938/230399849-d9e50cf6-cfcf-44a1-abae-60cd4eee9997.png)

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/pdf.clj:
```diff
@@ -544,17 +554,17 @@
-                           [(muodosta-pdf %)]))
-                sisalto))
-            #_[[:fo:block {:id "raportti-loppu"}]])))
+               (concat [;; Jos raportin tunnistetiedoissa on annettu :tietoja avaimella, näytetään ne alussa
```
Formatointi mennyt väärin tässä

solita-staskila replied:
Katsotaan tämä yhdessä

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,192 @@
+     [:varillinen-teksti {:arvo (pvm/pvm-aika (:toimenpiteet-aloitettu tiedot))}])))
+
+(defn- taulukko [{:keys [otsikot tiedot]}]
+  (let [rivit (into [] (map #(rivi-taulukolle %) tiedot))]
```
`mapv`-funktiolla saat tuloksen valmiiksi vektoriin lazyseq sijasta.
```suggestion
  (let [rivit (mapv #(rivi-taulukolle %) tiedot)]
```

solita-staskila replied:
Korjattu & testattu
https://github.com/finnishtransportagency/harja/pull/2824/commits/dafe947f3f58e3040f66fd053bf34e7f175987f4

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,192 @@
+        myohassa (filtteri-paalla-str (contains? vaikutukset :myohassa))
+        aiheutti-toimenpiteita (filtteri-paalla-str (contains? vaikutukset :aiheutti-toimenpiteita))
+
+        rivit (into []
```
`(into [] ...)` on tässä turha.
`(into [] [1 2 3 4])` on sama kuin `[1 2 3 4]`

solita-staskila replied:
Korjattu & testattu
https://github.com/finnishtransportagency/harja/pull/2824/commits/dafe947f3f58e3040f66fd053bf34e7f175987f4

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,192 @@
+        tiedot (:tiedot parametrit)
+        urakka (:urakka parametrit)
+        hallintayksikko (:hallintayksikko parametrit)
+        filtterit (:filtterit parametrit)
```
```suggestion
(defn suorita [_ _ {:keys [parametrit]}]
  (let [{:keys [tiedot urakka hallintayksikko filtterit]} parametrit
```

jpluomajoki-solita replied:
Grid näköjään tekee vähän hassuja juttuja parametreihin, sinne tulee nestattuna nuo parametrit. mallia
```
{:nimi "nimi"
 :parametrit {
   :nimi "nimi"
   :parametrit {
     :tiedot {...}
     :urakka 123
     ...jne}}}
```
Eli destrukturoidaan suorita-funktion parametreissa ensin kerran tuo parametrit, ja sitten voidaan vaikkapa letissä destrukturoida se toisen kerran. Voisi myös varmaan tehdä jotenkin tuossa parametreissa, mutta menisi sekavaksi.

solita-staskila replied:
Korjattu & testattu
https://github.com/finnishtransportagency/harja/pull/2824/commits/dafe947f3f58e3040f66fd053bf34e7f175987f4

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,192 @@
+        otsikko (if (and (not (:nimi urakka)) (not valittu-ely)) (str "Ilmoitukset, Koko maa") otsikko)
+
+        valinnat (:valinnat filtterit)
+        alkuaika (:valitetty-urakkaan-alkuaika valinnat)
```
Valinnat voidaan destrukturoida tässä letissä.
Joko:
```
        {alkuaika :valitetty-urakkaan-alkuaika
         tyypit :tyypit
         ilmoittaja-nimi :ilmoittaja-nimi
         hakuehto :hakuehto
         ilmoittaja-puhelin :ilmoittaja-puhelin
         toimenpiteet-aloitettu-vakioaikavali :toimenpiteet-aloitettu-vakioaikavali
         selite :selite
         tilat :tilat
         valitetty-urakkaan-vakioaikavali :valitetty-urakkaan-vakioaikavali
         tr-numero :tr-numero
         tunniste :tunniste
         vaikutukset :vaikutukset
         aloituskuittauksen-ajankohta :aloituskuittauksen-ajankohta} valinnat
```
tai sitten `:keys [valitetty-urakkaan-alkuaika jne...] valinnat`, joskin sitten täytyy käyttää samannimisiä muuttujia kuin avaimet on, eli alkuaika pitää vaihtaa sitten myöhemällä rivillä `valitetty-urakaan-alkuaika`

solita-staskila replied:
Korjattu & testattu
https://github.com/finnishtransportagency/harja/pull/2824/commits/dafe947f3f58e3040f66fd053bf34e7f175987f4

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -0,0 +1,192 @@
+                          :vaikutukset vaikutukset
+                          :aloituskuittauksen-ajankohta aloituskuittauksen-ajankohta
+                          :sheet-nimi "ilmoitukset"
+                          :otsikko "Filtterit"})
```
"Suodattimet" voisi olla ehkä hieman parempaa suomen kieltä

solita-staskila replied:
Tätä :otsikko avainta ei edes käytetä. Poistettu. 
Mutta rivi "Käytetyt filtterit" korjattu -> "Käytetyt suodattimet"

Korjattu https://github.com/finnishtransportagency/harja/pull/2824/commits/48b5c67b9ff0c8203358eb6ed71b21ea5ef5502e

jpluomajoki-solita commented on src/cljs/harja/ui/grid/perus.cljs:
```diff
@@ -366,8 +366,10 @@
-
+                                    true))
+              
+              raportin-napin-tyyli (if raporttivienti-lapinakyva? :button.nappi-toissijainen :button.nappi-ensisijainen)]
```
Tämän voi hoitaa myös luokalla, kts. ehdotus alempana

solita-staskila replied:
No mutta näinhän tämä on tehty? 
**:button.nappi-toissijainen** tekee buttonin jonka luokka on nappi-toissijainen.

Toinen vaihtoehto on {:class ""}, mutta lopputulos on sama.

Onko tässä niin väliä kumpaa syntaxia käyttää?

jpluomajoki-solita commented on src/cljs/harja/ui/grid/perus.cljs:
```diff
@@ -377,7 +379,7 @@
                              [:input {:type "hidden" :name "parametrit"
                                       :value ""}]
-                             [:button.nappi-ensisijainen
+                             [raportin-napin-tyyli
                               {:type "submit"
                                :class (when (or (not= idx (- (count valitut-raportin-vientimuodot) 1))
```
```suggestion
                             [:button
                              {:type "submit"
                               :class [(if raporttivienti-lapinakyva? "nappi-toissijainen" "nappi-ensisijainen")
                                 (when (or (not= idx (- (count valitut-raportin-vientimuodot) 1))]
```

jpluomajoki-solita commented on src/cljs/harja/views/ilmoitukset/tieliikenneilmoitukset.cljs:
```diff
@@ -32,6 +33,8 @@
   (:require-macros [cljs.core.async.macros :refer [go]]
                    [harja.tyokalut.ui :refer [for*]]))
 
+(def raportin-parametrit (atom {}))
```
Atomia ei tarvita, mieluummin pois. Kts. kommentit alempana.

solita-staskila replied:
Korjattu  https://github.com/finnishtransportagency/harja/pull/2824/commits/82d85da254993f2d4e89b47eb7ca49e5bfb4bde4

Mutta alkupvm ei tule enään raportin-yleiset-tiedot avaimeen yleiset.clj , tämä pitää korjata.

solita-staskila replied:
https://github.com/finnishtransportagency/harja/pull/2824/commits/11a2b2f5e4e0e0efe367931b1635a51b05694d22

jpluomajoki-solita commented on src/cljs/harja/views/ilmoitukset/tieliikenneilmoitukset.cljs:
```diff
@@ -207,7 +210,21 @@
+                                     :hallintayksikko @nav/valittu-hallintayksikko
+                                     :tiedot haetut-ilmoitukset
+                                     :filtterit @tiedot/ilmoitukset
+                                     :alkupvm alkuaika
```
alkuaika on aina nil, valintojen alustus on tainnut jäädä kesken, kts. ylempi kommentti

jpluomajoki-solita commented on src/cljs/harja/views/ilmoitukset/tieliikenneilmoitukset.cljs:
```diff
@@ -207,7 +210,21 @@
+                                       valinnat (:valinnat {})
+                                       alkuaika (:valitetty-urakkaan-alkuaika valinnat)]
+
+                                   (raportit/urakkaraportin-parametrit
```
Tätä ei tarvi kierrättää atomin kautta. voidaan käyttää tällaisenaan raporttiparametreinä.

jpluomajoki-solita replied:
Esim
```
        raportin-parametrit (raportit/urakkaraportin-parametrit
                              (:id @nav/valittu-urakka)
                              :ilmoitukset-raportti
                              {:urakka @nav/valittu-urakka
                               :hallintayksikko @nav/valittu-hallintayksikko
                               :tiedot haetut-ilmoitukset
                               :filtterit @tiedot/ilmoitukset
                               :alkupvm (:aloituskuittauksen-ajankohta valinnat-nyt)
                               :urakkatyyppi (:tyyppi @nav/valittu-urakka)})

```

jpluomajoki-solita commented on src/cljs/harja/views/ilmoitukset/tieliikenneilmoitukset.cljs:
```diff
@@ -242,7 +258,11 @@
+        
+        :raporttivienti #{:excel :pdf}
+        :raporttivienti-lapinakyva? true
+        :raporttiparametrit @raportin-parametrit}
```
Ja tässä voi sitten käyttää atomin sijasta letissä määriteltyä muuttujaa tai passata suoraan, esim.
```
        :raporttiparametrit (raportit/urakkaraportin-parametrit
                              (:id @nav/valittu-urakka)
                              :ilmoitukset-raportti
                              {:urakka @nav/valittu-urakka
                               :hallintayksikko @nav/valittu-hallintayksikko
                               :tiedot haetut-ilmoitukset
                               :filtterit @tiedot/ilmoitukset
                               :alkupvm (:aloituskuittauksen-ajankohta valinnat-nyt)
                               :urakkatyyppi (:tyyppi @nav/valittu-urakka)})}

```

jpluomajoki-solita commented on test/clj/harja/palvelin/raportointi/ilmoitukset_raportti_test.clj:
```diff
@@ -0,0 +1,65 @@
+            [harja.palvelin.palvelut.raportit :as raportit]
+            [harja.palvelin.palvelut.toimenpidekoodit :refer :all]
+            [harja.palvelin.palvelut.urakat :refer :all]
+            [harja.palvelin.raportointi :refer [suorita-raportti]]))
```
duplikaatti-require

solita-staskila replied:
https://github.com/finnishtransportagency/harja/pull/2824/commits/ff1a1c69bcd9845d466941144aa29d3fec35a1b3

jpluomajoki-solita commented on test/clj/harja/palvelin/raportointi/ilmoitukset_raportti_test.clj:
```diff
@@ -0,0 +1,65 @@
+            [harja.palvelin.komponentit.pdf-vienti :as pdf-vienti]
+            [harja.palvelin.komponentit.http-palvelin :as palvelin]
+            [harja.palvelin.raportointi :as raportointi]
+            [harja.palvelin.palvelut.raportit :as raportit]
```
```suggestion
            [harja.palvelin.palvelut.raportit :as raportit :refer [suorita-raportti]]
```

jpluomajoki-solita reviewed (changes requested):
Näyttää jo paljon paremmalta. Kommenteissa vielä muutamat siistimisehdotukset. Niiden jälkeen pitäisi olla hyvä.

solita-staskila committed:
Korjattu tuotekohtainen error

Jos tuotekohtaiseen laskutusyhteenvetoon yrittää mennä raportti- näkymän kautta, tämä heittää jostain syystä nullpointer exception
Tämä error korjattu.

solita-staskila committed:
Koodin siistintä

solita-staskila committed:
Korjattu suomennos

solita-staskila committed:
Turha atom pois

solita-staskila committed:
Duplikaatti require

solita-staskila committed:
Testi korjattu, try catch pois

solita-staskila committed:
Formatointi

solita-staskila committed:
Pävämäärä parametrit korjattu

solita-staskila committed:
Merge branch 'develop' into VHAR-7376_ilmoitukset_raportointi

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/laskutusyhteenveto_tuotekohtainen.clj:
```diff
@@ -80,18 +80,22 @@
+                                 (when (raha-arvo-olemassa? (data :hj_hoitovuoden_paattaminen_kattohinnan_ylitys_laskutettu))
+                                   (rivi-taulukolle data kyseessa-kk-vali? "Hoitovuoden päättäminen / Urakoitsija maksaa kattohinnan ylityksestä"
+                                                    :hj_hoitovuoden_paattaminen_kattohinnan_ylitys_laskutettu :hj_hoitovuoden_paattaminen_kattohinnan_ylitys_laskutetaan false))
+                                 (catch Throwable t
```
Mistä tässä blokissa edes voi nousta virhe? Lisäksi liian laaja catch ei ole hyvien käytäntöjen mukaista.

solita-staskila replied:
https://github.com/finnishtransportagency/harja/pull/2824/commits/4430c3ac4f260caa3c7d09940b217c5ba84011d9

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/laskutusyhteenveto_tuotekohtainen.clj:
```diff
@@ -80,18 +80,22 @@
-                                                  :hj_hoitovuoden_paattaminen_kattohinnan_ylitys_laskutettu :hj_hoitovuoden_paattaminen_kattohinnan_ylitys_laskutetaan false))
-
+                               (try
+                                 (when (raha-arvo-olemassa? (data :hj_hoitovuoden_paattaminen_tavoitepalkkio_laskutettu))
```
Pikkuasia: `data` on ilmeisesti map. Tyypillisempää olisi kutsua `(:avain map)` vs. `(map :avain)`. Funktionaalisesti sama asia, mutta näyttää omaan silmään hieman hassulta näin päin tehtynä

solita-staskila replied:
https://github.com/finnishtransportagency/harja/pull/2824/commits/4430c3ac4f260caa3c7d09940b217c5ba84011d9

jpluomajoki-solita commented on src/clj/harja/palvelin/raportointi/raportit/ilmoitukset.clj:
```diff
@@ -38,7 +38,7 @@
 
 (defn- taulukko [{:keys [otsikot tiedot]}]
-  (let [rivit (into [] (map #(rivi-taulukolle %) tiedot))]
+  (let [rivit (mapv #(rivi-taulukolle %) tiedot)]
```
```suggestion
  (let [rivit (mapv rivi-taulukolle tiedot)]
```
Ei tarvitse käyttää anonyymiä funktiota, kun se ottaa vain yhden parametrin

solita-staskila replied:
https://github.com/finnishtransportagency/harja/pull/2824/commits/f121c208cc571bae8a74e3f972b719bb41080906

jpluomajoki-solita reviewed (commented):
Muutama pieni koodityyliasia, suurimpana vähän vaarallisen näköinen try-catch

solita-staskila committed:
Korjattu virhe hakemalla data oikein + formatoinnit

solita-staskila committed:
Turha funktio mappaus pois

solita-staskila committed:
Merge branch 'develop' into VHAR-7376_ilmoitukset_raportointi
